### PR TITLE
Create exercise implemented (Pending video handling and optimization)

### PR DESCRIPTION
**Pull request by EvilDao-RG**
With this update, when a user tries to add an exercise to a routine, they have an option to create their own routine. They must select the body zone their exercise targets and add a title and caption to it if they want to. They can upload a photo and a video with it and the exercise is successfully saved to Parse.
All API requests are remained centralized in ParseAPIManager and there are basic validations to avoid incorrect or problematic uploads.
Also, a protocol is added to the view so that when a user saves the exercise, the exercise list could be updated without having to reload everything

**Issues**
Parse supports uploads of up to 20MB so it's almost impossible to upload photos and videos at the same time. So maybe I will upload the video to a different table inside the API or get another service that could store those videos

**Upload example**
ezgif-4-b731753042

Default input when caption and title are empty
Screen Shot 2022-07-14 at 9 55 51 AM

Alert when no body zone is selected
simulator_screenshot_649DDC32-3B46-48C9-B99A-E74ACB50E6FA

Default image when no image is uploaded
defaultImage

### CODE REVIEW

File: GymBuddies/APIManagers/ParseAPIManager.m [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
++ (void)saveExercise:(Exercise *)exercise
+           completion:(ParseManagerCreateCompletionBlock) completion {
+    Exercise *newExercise = [Exercise new];
+    newExercise.title = exercise.title;
+    newExercise.image = exercise.image;
+    newExercise.caption = exercise.caption;
+    newExercise.author = exercise.author;
+    newExercise.video = exercise.video;
+    newExercise.bodyZoneTag = exercise.bodyZoneTag;
```
[the remark this answers is not in the document]
Comment (radhikadhomse) [reply GymBuddies/APIManagers/ParseAPIManager.m:92]: or you may even add a copy method into your Exercise class so this copy over can be done internally

File: GymBuddies/ViewControllers/CreateExerciseViewController.m [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        PFFileObject *video = [ParseAPIManager getPFFileFromURL:urlVideo];
+        self.exerciseVideo = video;
+    } else {
+        self.imagePreview.image = [info objectForKey:UIImagePickerControllerOriginalImage];
+        PFFileObject *image = [ParseAPIManager getPFFileFromImage:self.imagePreview.image];
+        self.exerciseImage = image;
+        [self dismissViewControllerAnimated:YES completion:nil];
+    }
+    /*self.imagePreview.image = [info objectForKey:UIImagePickerControllerOriginalImage];
+    [self dismissViewControllerAnimated:YES completion:nil];*/
```
Comment (radhikadhomse) [inline GymBuddies/ViewControllers/CreateExerciseViewController.m:173]: delete the commentS?

File: GymBuddies/ViewControllers/CreateExerciseViewController.m [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+#pragma mark -Alerts
+
+-(void)_failedFetchingAlert:(NSString *)message{
+    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Error fetching data" message:message preferredStyle:UIAlertControllerStyleAlert];
+    
+    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
+    [alert addAction:okAction];
+    
+    [self presentViewController:alert animated:YES completion:nil];
```
Comment (radhikadhomse) [inline GymBuddies/ViewControllers/CreateExerciseViewController.m:185]: how about extracting all this into a helper and just passing in the differences (seems just the title)

File: GymBuddies/ViewControllers/LoginViewController.m [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+
+-(void)_wrongCredentialsAlert:(NSString *)message{
+    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Wrong username or password" message:message preferredStyle:UIAlertControllerStyleAlert];
+    
+    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
+    [alert addAction:okAction];
+
+    [self presentViewController:alert animated:YES completion:nil];
```
Comment (radhikadhomse) [inline GymBuddies/ViewControllers/LoginViewController.m:69]: if you end up extracting the alert code into helper, itd be usable here too!

File: GymBuddies/ViewControllers/RegisterViewController.m [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    
+- (BOOL)_lookForEmptyFields:(NSString *)username
+                      email:(NSString *)email
+                   password:(NSString *)password
+            confirmPassword:(NSString *)confirmPassword {
+    
+    return (username.length == 0 || password.length == 0 || confirmPassword == 0 || email.length == 0);
+
+}
+
```
Comment (radhikadhomse) [inline GymBuddies/ViewControllers/RegisterViewController.m:81]: not a requirement but you can consider just making a helper that takes in a list of strings and returns true if ANY are empty

File: GymBuddies/ViewControllers/RegisterViewController.m [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+#pragma mark - Alerts
+
+-(void)_emptyFieldAlert{
+    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Empty field" message:@"There is one or more empty fields" preferredStyle:UIAlertControllerStyleAlert];
+    
+    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
+    [alert addAction:okAction];
+    
+    [self presentViewController:alert animated:YES completion:nil];
```
Comment (radhikadhomse) [inline GymBuddies/ViewControllers/RegisterViewController.m:99]: similar to the above, use a nice  helper so we can deduplicate the code

File: GymBuddies/AppCommonResources/ColorPallet.h [hunk 7]
```diff
@@ -0,0 +1,12 @@
+//
```
[the remark this answers is not in the document]
Comment (EvilDao-RG) [reply GymBuddies/AppCommonResources/ColorPallet.h:1]: Maybe I should delete it for now. I'm not applying colors at his moment and now I don't know if I'll need it

File: GymBuddies/APIManagers/ParseAPIManager.h [hunk 8]
```diff
@@ -0,0 +1,41 @@
+//
+//  ParseManager.h
```
[the remark this answers is not in the document]
Comment (EvilDao-RG) [reply GymBuddies/APIManagers/ParseAPIManager.h:2]: GymBuddiesDBManager?
GymBuddiesAPIManager?
GymBuddiesNetworkManager?

File: GymBuddies/ViewControllers/CreateExerciseViewController.m [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+        } else {
+            self.bodyZones = elements;
+            [self.bodyZoneCollectionView reloadData];
+        }
+    }];
+}
+
+
+- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView{
+    return 1;
```
[the remark this answers is not in the document]
Comment (EvilDao-RG) [reply GymBuddies/ViewControllers/CreateExerciseViewController.m:106]: True. And I think that's the default value as well

File: GymBuddies/ViewControllers/CreateExerciseViewController.m (hunk 3 above)
Comment (EvilDao-RG) [reply GymBuddies/ViewControllers/CreateExerciseViewController.m:185]: Ok. I'll try that. Maybe I'll change the name of the CommonValidations.h to include the alert methods

**Pull request merged**